libs/algo/src/lib.rs

```diff
@@ -0,0 +1,105 @@
+use pyo3::prelude::{pymodule, PyModule, PyResult, Python};
+use numpy::{PyArray4, PyArray3, ToPyArray};
+use ndarray::{array, Array, ArrayBase, ArrayView, ArrayViewMut, Axis, Ix1, Ix2, Ix3, Ix4};
+
+
+fn geomedian<'a>(in_array: ArrayView<'a, f32, Ix4>, max_iter: usize, eps: f32) -> Array<f32, Ix3> {
+    
+    let shape = in_array.shape();
+    let rows = shape[0];
+    let columns = shape[1];
+    let bands = shape[2];
+
+    let mut out_array: Array<f32, Ix3> = ArrayBase::zeros([rows, columns, bands]);
+
+    for row in 0..shape[0] {
+        for column in 0..shape[1] {
+            geomedian_pixel(in_array, &mut out_array, row, column, max_iter, eps);
+        }
+    }
+
+    out_array
+}
+
+
+fn geomedian_pixel<'a>(
+    in_array: ArrayView<'a, f32, Ix4>,
+    out_array: &mut Array<f32, Ix3>,
+    row: usize,
+    column: usize,
+    max_iter: usize, 
+    eps: f32
+) { 
+
+    let bands = in_array.shape()[2];
+    let time_steps = in_array.shape()[3];
+    
+    let mut valid_time_steps: Vec<usize> = Vec::with_capacity(time_steps);
+    for t in 0..time_steps {
```

Comment:
Rather than keeping indexes of valid time slices, it's probably better to extract actual data into temp array (re-used between pixels columns)
- better data cache performance independent of source data memory layout
- allows for arbitrary pre-processing steps (scale normalization or uint16 -> float32 conversion)

Reply:
I wasn't sure whether this would be worth the trade-off of copying the data around. If most of the measurements are valid the cache performance would be pretty good anyway, but then again I have no idea what this data is like in practice! Roughly how much of the data is valid?

Reply:
About 50% I'd say, but it also matters how it is stored